### PR TITLE
Async usage with new websockets and http async transports using asyncio

### PR DESCRIPTION
**Pull request by leszekhanusz**
Necessitate Python 3.6 and asyncio

### New Client
* can be used as before synchronously
```python
client.execute(query)
```
* but can also be used asynchronously
```python
async with Client(...) as session:
    await session.execute(query)
```
* supports existing sync http transport (requests) (only synchronously)
* supports also new async transports (websockets and aiohttp)
* Supports subscriptions with the websockets transport
* synchronously with an infinite python generator
```python
for result in client.subscribe(query):
    ... result is a dict here
```

* OR asynchronously with an async generator.
```python
async with Client(...) as session:
    async for result in session.subscribe(query):
        ... result is a dict here
```
* for async transports, fetching schema from transport is done when executing the first query

### New Websockets async transport

Allows a connection to an Apollo GraphQL websocket endpoint using the [apollo subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) protocol

Require the great [websockets library](https://websockets.readthedocs.io/en/stable/intro.html)

Supports:

* queries, mutations AND subscriptions
* SSL encryption
* multiple queries using the same connection
* multiple subscription in parallel on the same connection using asyncio tasks

### New HTTP async transport

Supports thanks to the [aiohttp library](https://docs.aiohttp.org/en/stable/):
* http headers
* http cookies
* Basic auth
* SSL

### gql-cli

* Added gql-cli script to execute graphql queries from the command line to an http or websocket endpoint.

### CODE REVIEW

**Comment by yiblet** [comment]
thanks for doing this work, I was looking exactly for these features!

**Comment by KingDarBoja** [comment]
Been busy due to real-life stuff but will look into this PR as soon as possible.

File: setup.py [hunk 1]
```diff
@@ -6,7 +6,8 @@
     'six>=1.10.0',
     'graphql-core>=2,<3',
     'promise>=2.0,<3',
-    'requests>=2.12,<3'
+    'requests>=2.12,<3',
+    'websockets>=8.1,<9'
```
Comment (KingDarBoja) [inline setup.py:10]: From pypi, websockets support Python 3.6.1 and higher. This one should be appended to the `install_requires` list. See line 20 for an example with `pytest-asyncio`.

Comment (leszekhanusz) [reply setup.py]: Done. Now tests are passing. Will need to add tests to raise the coverage now.

Comment (leszekhanusz) [reply setup.py]: Also now that I understand asyncio a little better I think it is best to use Futures intead of Queues for the listeners. Will need to change that also so don't accept this PR just yet.

Comment (KingDarBoja) [reply setup.py]: No problem mate :)

**Comment by KingDarBoja** [comment]
This is looking great, of course requires several reviews from the graphql-python team but glad to see this effort 🚀

File: setup.py (hunk 1 above)
Comment (leszekhanusz) [reply setup.py]: In fact queues are really needed for slow consumers ...

**Comment by maeglindeveloper** [comment]
Hi everyone, is there any example for that new feature :) !?

**Comment by leszekhanusz** [comment]
> Hi everyone, is there any example for that new feature :) !?
> I see also that the merging is blocked, do you have an ETA of release ?
> Thanks for the good work ! +1

There are examples added in the README.md
You can see them here [https://github.com/leszekhanusz/gql](https://github.com/leszekhanusz/gql) in the "Websockets transport with asyncio" section

**Comment by leszekhanusz** [comment]
Note: should I rebase my pull request and squash the small fix commits ?

This would avoid the merge commit but would rewrite history in this PR

**Comment by maeglindeveloper** [comment]
@leszekhanusz, the best way is to have it review no ?  @syrusakbary, wdyt ?
Then new release can be done ?

**Comment by maeglindeveloper** [comment]
Any updates on this ? :

**Comment by KingDarBoja** [comment]
@leszekhanusz don't worry about that, at the end this PR will be squashed into a single commit to keep things clean. I will review this PR after discussing with **graphql-python** folks about the next release (v3). 

Please read [Cito comment](https://github.com/graphql-python/gql/issues/62#issuecomment-619544513) about what I am talking about.

**Comment by maeglindeveloper** [comment]
Keep us update :) ! I'm looking forward to use that new feature.
Great job guys

**Comment by sneko** [comment]
Awesome @leszekhanusz! That's cool this library will benefit from this, congrats!

Just curious, may I missed something but will we be able to deal with the initial payload at init?

Thank you,

**Comment by leszekhanusz** [comment]
> Just curious, may I missed something but will we be able to deal with the initial payload at init?
Do you mean the payload we are sending in the connection_init message ? Currently this payload is empty but we could add a parameter init_payload to the transport for it.

**Comment by sneko** [comment]
Yeah I was thinking about this.

In another project using Javascript I had to use it to pass some init parameters but also passing the access token to allow or not establishing the connection. At this time I had no other choice for the access token (not even using a header I remember).

**Comment by sneko** [comment]
Woow, such reactivity! Thank you for your work. I hope it will pass the review to try all of this 👍 🎉

**Comment by Cito** [comment]
Thanks again for this contribution. After creating a v2.x branch for the legacy support, and declaring the master to be the modern v3 branch, I merged this into master, and after fixing a small merge problem, tests are green. I will do some clean-up and modernization work in the master branch now.

**Pull request merged**